analyses/mwi/dryspells_trigger/exploration/ecmwf_realtime_test.md

```diff
@@ -0,0 +1,256 @@
+config=Config()
+parameters = config.parameters(country_iso3)
+
+country_data_raw_dir = os.path.join(config.DATA_DIR,config.PUBLIC_DIR, config.RAW_DIR,country_iso3)
+adm1_bound_path=os.path.join(country_data_raw_dir,config.SHAPEFILE_DIR,parameters["path_admin1_shp"])
+ecmwf_realtime_dir = Path(config.DATA_DIR) / config.PRIVATE_DIR / config.RAW_DIR / "glb" / "ecmwf"
+```
+
+```python
+gdf_adm1=gpd.read_file(adm1_bound_path)
+```
+
+We select all files that contain the seasonal forecast with monthly mean, i.e. that Have T4L in their filename
+
+```python
+filepath_list = [
+        filename
+        for filename in ecmwf_realtime_dir.iterdir()
+        #T4L indicates the seasonal forecast with the monthly mean
+        #see here for a bit more explanation https://confluence.ecmwf.int/pages/viewpage.action?pageId=111155348
+        #without the 1 you also get .idx files which we don't want but not sure if this is the best method to select
+        if fnmatch.fnmatch(filename, '*T4L*1') 
+    ]
+```
+
+```python
+filepath_list
+```
+
+We first load one file without any concatenation and processing to understand the fileformat.  
+
+The files are very cryptic but they contain a dataType which can be "em" or "fcmean". As far as I understand "em" is the mean across all ensembles while "fcmean" is the value per ensemble 
+
+Then there also is the "numberOfPoints". This has to do with the geographical area, since we receive the forecast for several geographical areas. 
+Just by testing and looking at the latitude/longitude range I figured 384 is the one matching Malawi but no idea if there is a better way to set this
+
+```python
+xr.load_dataset(filepath_list[0],engine="cfgrib",filter_by_keys={'numberOfPoints': 384, 'dataType': 'fcmean'})
+```
+
+Next we load all files at once
+
+```python
+def _preprocess_monthly_mean_dataset(ds_date: xr.Dataset):
+        
+         return (
+            ds_date.assign_coords(
+                {
+                    #get the difference between the valid time month and the publication month 
+                    #valid time is one month ahead of the start of the actual valid time
+                    #so a leadtime of 1 means that the forecast is forecasting the same month as the publication month
+                    "step": (ds_date.valid_time.dt.month-ds_date.time.dt.month)+12*(ds_date.valid_time.dt.year-ds_date.time.dt.year)
+                }
+            )
+            .expand_dims("time")
+            #valid time is one month ahead of the start of the actual valid time --> only confusing so drop
+            #surface is empty
+             .drop_vars(["valid_time","surface"])
+        )
+```
+
+```python
+#i would think setting concat_dim=["time","step"] makes more sense 
+#but get an error "concat_dims has length 2 but the datasets passed are nested in a 1-dimensional structure"
+#it seems to work thought when using concat_dim="time" but would have to test once we have data from several dates.. 
```

Comment:
I think this error makes sense because `time` and `step` are a single dimension

Reply:
I don't think I fully understand. They are both individual dimensions right? Aah or wait is it the difference between coords and dim? 
From this error, do you then understand how we could merge the different files based on both `time` and `step`?

Reply:
I was just using dimension in the general sense (not the xarray definition). What I mean is that they are interchangeable, rather than orthogonal dimensions, so you really only need to specify one and the other will be known. I hope that makes sense.

Reply:
Not sure I understand. Do you mean that individually they are not unique but together they are, so that is automatically determined if one of them is passed?

Reply:
Yes I think so. One of them is just a simple transformation of the other. You could remove one and the dataset would still be fully defined.

Reply:
>You could remove one and the dataset would still be fully defined. 

Hmm but I wouldn't agree with that. `time` only contains one value per file, namely the publishing month. While `step` contains 7 values indicating the leadtimes. If we would remove one of the two, we couldn't fully define the dataset.

Reply:
Ah shoot, I'm so sorry Tinka, I was mixing up (what are now) `time` and `verifying_time`. My arguments were all based on `verifying_time`, I hope that makes more sense.

In the case of `time`, I think if you had other files where `time` would be different, it would need to be included in the list of `concat_dim`. However, since there is only one value for `time` then it doesn't treat it as an extra dimension.

Reply:
Yes that might be the case